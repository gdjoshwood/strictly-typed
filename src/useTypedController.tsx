import React from 'react';
import { useFormContext, Controller, Control } from 'react-hook-form';
import { formatName } from './formatName';
import {
  DeepPath,
  FieldValuesFromControl,
  Options,
  ControllerProps,
} from './types';

export const useTypedController = <
  TFieldValues extends FieldValuesFromControl<TControl>,
  TControl extends Control = Control
>({
  control: controlFromOptions,
}: Options<TControl>) => {
  const methods = useFormContext<TFieldValues>();
  const control = controlFromOptions || methods.control;

  const TypedController = React.useCallback(
    <
      TFieldName extends DeepPath<TFieldValues, TFieldName>,
      TAs extends 'input' | 'select' | 'textarea'
    >({
      name,
      as,
      rules,
      ...rest
    }: ControllerProps<TFieldValues, TFieldName, TAs>) => {
      const formattedName = formatName(name as any);
      const { register } = control;
      if (as) {
        return React.createElement(as, {
          name: formattedName,
          ref: rules ? register(rules as any) : register,
          ...rest,
        });
      }
      return (
        <Controller
          name={formattedName as any}
          control={control as any}
          rules={rules}
          {...rest}
        />
      );
    },
    [],
  );

  return TypedController;
};
