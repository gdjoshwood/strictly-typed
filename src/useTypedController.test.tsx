import React from 'react';
import { Control } from 'react-hook-form';
import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import { useTypedController } from './useTypedController';

export const reconfigureControl = (
  controlOverrides: Partial<Control> = {},
): Control => ({
  unmountFieldsStateRef: {
    current: {},
  },
  defaultValuesRef: {
    current: {},
  },
  isWatchAllRef: {
    current: false,
  },
  validFieldsRef: {
    current: new Set(),
  },
  fieldsWithValidationRef: {
    current: new Set(),
  },
  fieldArrayDefaultValues: {
    current: {},
  },
  watchFieldsRef: {
    current: new Set(),
  },
  watchFieldsHookRef: {
    current: {},
  },
  watchFieldsHookRenderRef: {
    current: {},
  },
  watchInternal: jest.fn(),
  validateResolver: jest.fn(),
  setValue: jest.fn(),
  getValues: jest.fn(),
  register: jest.fn(),
  unregister: jest.fn(),
  trigger: jest.fn(),
  removeFieldEventListener: jest.fn(),
  mode: {
    isOnSubmit: false,
    isOnBlur: false,
    isOnChange: false,
    isOnTouch: false,
    isOnAll: false,
  },
  reValidateMode: {
    isReValidateOnBlur: false,
    isReValidateOnChange: false,
  },
  formStateRef: {
    current: {
      errors: {},
      isDirty: false,
      isSubmitted: false,
      dirtyFields: {},
      submitCount: 0,
      touched: {},
      isSubmitting: false,
      isValid: false,
    },
  },
  fieldsRef: {
    current: {},
  },
  resetFieldArrayFunctionRef: {
    current: {},
  },
  fieldArrayNamesRef: {
    current: new Set<string>(),
  },
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  updateFormState: () => {},
  readFormStateRef: {
    current: {
      isDirty: true,
      errors: true,
      isSubmitted: false,
      submitCount: false,
      touched: false,
      isSubmitting: false,
      isValid: false,
      dirtyFields: false,
    },
  },
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  renderWatchedInputs: () => {},
  ...controlOverrides,
});

type FormValues = {
  flat: string;
  nested: {
    object: { test: string };
    array: { test: string }[];
  };
};

describe('useTypedField', () => {
  it('should render correctly when as is input and name is string', () => {
    const control = reconfigureControl();
    const { result } = renderHook(() =>
      useTypedController<FormValues>({
        control,
      }),
    );
    const TypedController = result.current;

    const { asFragment } = render(
      <TypedController as="input" name="flat" defaultValue="test" />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly when name is string and render is textarea', () => {
    const control = reconfigureControl();
    const { result } = renderHook(() =>
      useTypedController<FormValues>({
        control,
      }),
    );
    const TypedController = result.current;

    const { asFragment } = render(
      <TypedController
        name="flat"
        defaultValue="test"
        render={(props) => <textarea {...props} />}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly when as is textarea and name is array', () => {
    const control = reconfigureControl();
    const { result } = renderHook(() =>
      useTypedController<FormValues>({
        control,
      }),
    );
    const TypedController = result.current;

    const { asFragment } = render(
      <TypedController
        as="textarea"
        name={['nested', 'object', 'test']}
        defaultValue="test"
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly when name is array and render is input', () => {
    const control = reconfigureControl();
    const { result } = renderHook(() =>
      useTypedController<FormValues>({
        control,
      }),
    );
    const TypedController = result.current;

    const { asFragment } = render(
      <TypedController
        name={['nested', 'array', 0, 'test']}
        defaultValue="test"
        render={(props) => <input {...props} />}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly when name is string', () => {
    const control = reconfigureControl();
    const { result } = renderHook(() =>
      useTypedController<FormValues>({
        control,
      }),
    );
    const TypedController = result.current;

    render(
      <TypedController
        name="flat"
        defaultValue=""
        render={(props) => <input {...props} />}
      />,
    );

    expect(control.register).toHaveBeenCalledWith(
      {
        focus: undefined,
        name: 'flat',
      },
      undefined,
    );
  });

  it('should format name correctly when name is array (dot syntax)', () => {
    const control = reconfigureControl();
    const { result } = renderHook(() =>
      useTypedController<FormValues>({
        control,
      }),
    );
    const TypedController = result.current;

    render(
      <TypedController
        name={['nested', 'object', 'test']}
        defaultValue=""
        render={(props) => <input {...props} />}
      />,
    );

    expect(control.register).toHaveBeenCalledWith(
      {
        focus: undefined,
        name: 'nested.object.test',
      },
      undefined,
    );
  });

  it('should format name correctly when name is array (dot-bracket syntax)', () => {
    const control = reconfigureControl();
    const { result } = renderHook(() =>
      useTypedController<FormValues>({
        control,
      }),
    );
    const TypedController = result.current;

    render(
      <TypedController
        name={['nested', 'array', 0, 'test']}
        defaultValue=""
        render={(props) => <input {...props} />}
      />,
    );

    expect(control.register).toHaveBeenCalledWith(
      {
        focus: undefined,
        name: 'nested.array[0].test',
      },
      undefined,
    );
  });
});
