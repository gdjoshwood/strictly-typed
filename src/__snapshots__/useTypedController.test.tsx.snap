// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTypedField should render correctly when as is input and name is string 1`] = `
<DocumentFragment>
  <input
    value="test"
  />
</DocumentFragment>
`;

exports[`useTypedField should render correctly when as is textarea and name is array 1`] = `
<DocumentFragment>
  <textarea>
    test
  </textarea>
</DocumentFragment>
`;

exports[`useTypedField should render correctly when name is array and render is input 1`] = `
<DocumentFragment>
  <input
    value="test"
  />
</DocumentFragment>
`;

exports[`useTypedField should render correctly when name is string and render is textarea 1`] = `
<DocumentFragment>
  <textarea>
    test
  </textarea>
</DocumentFragment>
`;
